pkg/util/connectionpool/grpc/pool.go

```diff
@@ -0,0 +1,311 @@
+		spec.MaxConnsPerHost = defaultMaxIdleConnsPerHost
+	}
+	if spec.DialOptions == nil {
+		spec.DialOptions = defaultDialOpts
+	}
+
+	logger.Infof("created grpc connection pool with factory and spec %+v", spec)
+	return &Pool{
+		spec:       spec,
+		connsMutex: &sync.Mutex{},
+		conns:      map[string]*segmentConn{},
+		factory:    factory,
+	}
+}
+
+func (s *segmentConn) buildConnectionWithChannel(factory connectionpool.CreateConnFactory, spec *Spec, errCh chan<- error) {
+	s.Lock()
+	defer s.Unlock()
+	waited := int(atomic.LoadInt32(&s.waitedNum))
+	if waited <= len(s.connCh) || len(s.connCh) == cap(s.connCh) {
+		return
+	}
+	// it means all conn in used, so we pick random conn
+	if len(s.connMap) == int(spec.MaxConnsPerHost) {
+		for c := range s.connMap {
+			if c.GetState() == connectivity.Shutdown {
+				delete(s.connMap, c)
+				continue
+			}
+			s.connCh <- &Connection{ClientConn: c, segment: s, needBack: false}
+			return
+		}
+	}
+
+	for i := 0; i < int(spec.InitConnNum); i++ {
+		rawConn, err := buildConnection(s.addr, factory, spec)
+		if err != nil {
+			errCh <- status.Errorf(codes.Internal, "couldn't create connection for addr %s, cause: %s", s.addr, err.Error())
+			logger.Infof("not create connection for addr %s, cause: %s", s.addr, err.Error())
+			return
+		}
+		c := &Connection{
+			segment:    s,
+			ClientConn: rawConn,
+			needBack:   true,
+		}
+		s.connCh <- c
+		s.connMap[rawConn] = nil
+		// other goroutine return conn
+		if len(s.connMap) >= int(spec.MaxConnsPerHost) {
+			break
+		}
+	}
+}
+
+func (m *Pool) Get(addr string) (interface{}, error) {
+	timeout, cancelFunc := context.WithTimeout(context.Background(), m.spec.BorrowTimeout)
+	defer cancelFunc()
+	segment := m.conns[addr]
+	if segment == nil {
+		m.connsMutex.Lock()
+		if m.conns[addr] == nil {
+			m.conns[addr] = &segmentConn{
+				connCh:    make(chan *Connection, m.spec.MaxConnsPerHost),
+				connMap:   make(map[*grpc.ClientConn]interface{}, m.spec.MaxConnsPerHost),
+				Mutex:     &sync.Mutex{},
+				waitedNum: 0,
+				addr:      addr,
+			}
+		}
+
+		segment = m.conns[addr]
+		m.connsMutex.Unlock()
+	}
+
+	atomic.AddInt32(&segment.waitedNum, 1)
+	defer atomic.AddInt32(&segment.waitedNum, -1)
+	errCh := make(chan error, 1)
+	go segment.buildConnectionWithChannel(m.factory, m.spec, errCh)
+	for {
+		select {
+		case <-timeout.Done():
+			return nil, timeout.Err()
+		case c := <-segment.connCh:
+			if c.isAvailable() {
+				return c, nil
+			}
+			c.destroyConn()
+			go segment.buildConnectionWithChannel(m.factory, m.spec, errCh)
+		case err := <-errCh:
+			return nil, err
+		}
+	}
+}
+
+// ReleaseConn grpc share connection, so no need release
+func (m *Pool) ReleaseConn(i interface{}) {
+	if conn, ok := i.(*Connection); ok {
+		if !conn.isAvailable() {
+			conn.destroyConn()
+		} else {
+			conn.backChannel()
+		}
+	}
+}
+
+func (m *Pool) Close() {
+	if m.isDone {
+		return
+	}
+	m.connsMutex.Lock()
+	defer m.connsMutex.Unlock()
```

Comment:
in most cases, we should not use a lock in the cleanup functions, it is often useless or error prone.